janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/JanusGraphAbstractAuthenticator.java

```diff
@@ -0,0 +1,130 @@
+     */
+    public static final String CONFIG_CREDENTIAL_DB = "credentialDb";
+
+    /**
+     * Default username
+     */
+    public static final String CONFIG_DEFAULT_USER = "defaultUsername";
+
+    /**
+     * Default password for default username.
+     */
+    public static final String CONFIG_DEFAULT_PASSWORD = "defaultPassword";
+
+    protected CredentialGraph credentialStore;
+
+    @Override
+    public boolean requireAuthentication() {
+        return true;
+    }
+
+    @Override
+    abstract public SaslNegotiator newSaslNegotiator(final InetAddress remoteAddress);
+
+    @Override
+    abstract public SaslNegotiator newSaslNegotiator();
+
+    public CredentialGraph createCredentialGraph(JanusGraph graph) {
+        return new CredentialGraph(graph);
+    }
+
+    public JanusGraph openGraph(String conf) {
+        return JanusGraphFactory.open(conf);
+    }
+
+    public void setup(final Map<String,Object> config) {
+        logger.info("Initializing authentication with the {}", this.getClass().getName());
+        if (null == config) {
+            throw new IllegalArgumentException(String.format(
+                    "Could not configure a %s - provide a 'config' in the 'authentication' settings",
+                    this.getClass().getName()));
+        }
+
+        if (!config.containsKey(CONFIG_CREDENTIAL_DB)) {
+            throw new IllegalStateException(String.format(
+                    "Credential configuration missing the %s key that points to a graph config file or graph name", CONFIG_CREDENTIAL_DB));
+        }
+
+        final JanusGraph graph = openGraph(config.get(CONFIG_CREDENTIAL_DB).toString());
+        credentialStore = createCredentialGraph(graph);
+        graph.tx().rollback();
+        ManagementSystem mgmt = (ManagementSystem) graph.openManagement();
+        if (!mgmt.containsGraphIndex("byUsername")) {
+            final PropertyKey username = mgmt.makePropertyKey(PROPERTY_USERNAME).dataType(String.class).cardinality(Cardinality.SINGLE).make();
+            JanusGraphIndex index = mgmt.buildIndex("byUsername", Vertex.class).addKey(username).unique().buildCompositeIndex();
+            mgmt.commit();
+
+            mgmt = (ManagementSystem) graph.openManagement();
+            index = mgmt.getGraphIndex("byUsername");
+            if (!index.getIndexStatus(username).equals(SchemaStatus.ENABLED)) {
+                try {
+                    mgmt = (ManagementSystem) graph.openManagement();
+                    mgmt.updateIndex(mgmt.getGraphIndex("byUsername"), SchemaAction.REINDEX);
+                    mgmt.awaitGraphIndexStatus(graph, "byUsername").status(SchemaStatus.ENABLED).call();
+                } catch (InterruptedException rude) {
+                    throw new RuntimeException("Timed out waiting for byUsername index to be created on credential graph", rude);
+                }
+            }
+        }
+
+        if (credentialStore.countUsers() == 0) {
```

Comment:
This calls `hasLabel()` and is a potential huge performance hit when there are users populated. I'm not sure I agree with creating a default user.  But you could move this up as a part of creating the schema.

Reply:
There's a unique constraint on the index, I could try catch on the user creation instead of this check

Reply:
No, creating the default user is only good when creating the initial schema.  The default user should/would be removed when going to a deployed system.  You don't want to end up unwittingly recreate the default user every restart thus creating a security hole.  Maybe even attempting to create the default user should be a configuration - optional and false by default.